zigpy/application.py

```diff
@@ -462,6 +477,32 @@ def get_device(self, ieee=None, nwk=None):
 
         raise KeyError
 
+    def get_ephemeral_device(self, ieee, nwk) -> zigpy.device.Device:
+        """Get ephemeral device with currently unknown ieee or nwk."""
+
+        clear_expired = [d for d in self._ephemeral_devices if not d.expired]
```

Comment:
If an application holds a reference to an ephemeral device and the device "expires", this seems like it will cause unexpected behavior when zigpy creates a new clone.

What do you think about storing ephemeral devices in a [`weakref.WeakValueDictionary`](https://docs.python.org/3/library/weakref.html#weakref.WeakValueDictionary) (since there is no `WeakList`, maybe with just an incrementing key)?  That way, we can hold strong references to the devices in `_ephemeral_devices_refs` and expire them the way you're doing it here (or is `list.remove()` required?) but zigpy won't replace an ephemeral device with an identical clone when it's still in use.

Reply:
By application, do you mean ZHA or `ControllerApplication` ?  There could be trouble indeed if an ephemeral devices goes up all the way to ZHA. May be guard the origination of `device_initialized` event only for devices with fully discovered addresses? I haven't tested, but I would not expect it get there though, not until the device is fully initialized :thinking: 

Other than that, the `ControllerApplication` wouldn't store references to those devices anywhere, would it? except it could in topology ? https://github.com/zigpy/zigpy/blob/0c4b9a8a25c5eef5b008d66c2e56aae3f8d5e769/zigpy/neighbor.py#L31

Reply:
I need to play with the `WeakValueDictionary`, this sounds like a better approach here indeed.